### ISSUE TITLE
suggestion: providing a way to add custom labels on web pages on IE will be helpful

### ISSUE THREAD

**Issue by sumandogra** [nvaccess/nvda#5874]
At times there are web pages that have elements that do not have any labels given.  but one has to revisit such web pages and use such elements.  without the facility of adding custom labels it becomes difficult to use such web pages.  it is more important at professional places where an organization has designed their own work environment with such poorly accessible labelled web pages.

**Comment by ManshulBelani** [comment]
Hi,

I am working on  a solution for assigning and reading Custom Labels for webpages. The solution basically allows the user to assign a custom label to any field on the webpage. This label added by the user is added to a config file where the key of the key value pair is being provided by the HTML name attribute. 
For reading the assigned custom label, I have used the getName function of the MSHTML class. 
However these custom labels are being announced only if we are navigating between the webpage fields in the edit mode. While navigating in the browse mode, this information is not being announced by NVDA.
I tried adding the name attribute to the attributes map using addHTMLAttributeToMap macro being used for other similar attributes in mshtml.cpp file. However the info is still not being added to the virtual buffer.
Is there some way that this information can be added to the virtual buffer so that it gets announced in the browse mode as well.

**Comment by jcsteh** [comment]
As noted on our phone call, there's no way currently for you to tweak the actual content of the buffer. Also, browse mode reporting uses control fields, not NVDAObjects. You should be able to have the name spoken by doing something like this:

* Expose the attribute from the buffer using the macro you described.
* In virtualBuffers.MSHTML.MSHTMLTextInfo._normalizeControlField:
 1. Get the attribute you exposed. It'll probably be called something like IHTMLDOMNode::name.
 2. Map it to the name set by the user.
 3. Set that as an attribute called "name".
 4. Also set the attribute "alwaysReportName" to True.

**Comment by sumandogra** [comment]
I tested with the test build and made the following observations with the build:
1.	Please activate enter key in the edit field in the custom label dialog so that if one presses enter  once in the edit field, the custom labels are created.  One need not go to the ok button.
2.	Radio buttons in same group are assigned the same custom label.
3.	List box do not get the custom label dialog.
4.	Combo box does not accept the custom label.
5.	On buttons, first the label originally assigned by the web page are announced then the custom label is announced. Sometimes such labels are vague and should be particularly reassigned a name.  in such cases only the custom labels should be announced.
6.	Buttons that have submenus on them are not available for custom label and give ding error sound.  on using custom label command Please check at www.yahoo.com.
7.	Multiline edit fields also do not take the custom labels.


8.

**Comment by sumandogra** [comment]
scenario: The behavior should be uniform with arrows, tab and shift+tab, quick navigation keys.

On using quick navigation key and tab or shift+ tab only the custom label is being announced.  But on using arrows, the text of the undesired label that is present before the field also gets announced causing confusion.  The behavior should be same with arrows as well and only the custom label should be announced.

**Comment by ManshulBelani** [comment]
For adding and reading custom labels for links I tried setting the renderChildren=false after mapping the custom label to the contentString in mshtml.cpp. Still the customLabel is not being spoken by NVDA. 
What I am basically trying to do is that check the node type, if it is an anchor node and has an href attribute associated, the href attribute value of this node is matched to all entries of a map which contains all the values from the custom label configuration file. If this attribute value matches to either of the entries from the file, the corresponding custom label value from the file is mapped to the content string.I have applied logs and checked, the contentString is getting the correct value. However, this is probably not being added to the buffer. Even making renderChildren = false is not changing anything. Also the text value of the anchor node which was earlier being spoken becomes blank.

Could you please suggest some way for this to added to the buffer as well.

**Comment by ManshulBelani** [comment]
I was able to figure out the problem with this. By mistake, I had redeclared the variable contentString in my code and therefore its value was getting null outside the scope of the added code. The custom label is now being added to the buffer.

**Comment by ManshulBelani** [comment]
I was trying to add the custom label functionality for links and images and on testing I observed that when in the focus mode the href or src attribute is coming out to be accurate. However when in browse mode, sometimes this attribute returned is not the one of the object on which NVDA focus is there. This is creating some problems while assigning custom labels. The custom label value gets added to an incorrect key in the config file. Is there some way in which the attribute of the correct object (on which the NVDA focus is present) can be obtained when navigating in browse mode.

**Comment by Qchristensen** [comment]
Has anyone looked into this feature lately?  It is still commonly requested, for all browsers.

**Comment by Brian1Gaff** [comment]
It would be nice for many controls. I am sick of button button button on 
some sites, but the question is, is there sufficient difference available on 
the buttons code to allow one to tell the difference between them?



bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.

**Comment by Adriani90** [comment]
This function is commonly requested, not only for Browsers but also for every Kind of application on Windows. But this is not very simple to implement I guess. for this functionality it should be possible to customize the Information which Comes from the accessibility API. This is something which is probabllly not very welcome by Microsoft because this means we could manipulate the Information provided by the API. This can be interpreted as a security issue. I think detailed discussions with Microsoft are indeed required in this case.

**Comment by Adriani90** [comment]
@Qchristensen this may be a Point which NV Access could discuss with Microsoft at one of the regular conferences. What do you think?

**Comment by ruifontes** [comment]
In my opinion, we should not "customize the Information which Comes from 
the accessibility API", but create a NVDA type of dictionary to that 
information of the accessibility API to that page or app, and it will be 
used only on that page or app...

Rui Fontes



Às 14:48 de 17/12/2018, Adriani90 escreveu:

**Comment by LittleStar-VIP** [comment]
Yes, I would also want this function, and not just being able to label web page elements, but also any element in any application if possible.

**Comment by Cleversn** [comment]
I need this feature as well.

**Comment by seanbudd** [comment]
Closing in favour of #14821 and #17688
